repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -3,117 +3,153 @@
 import com.epam.esm.exception.DataException;
 import com.epam.esm.repository.SearchCriteria;
-import com.epam.esm.repository.builder.QueryBuilder;
-import com.epam.esm.repository.mapper.CertificateMapper;
 import com.epam.esm.repository.repositoryinterfaces.CertificateRepository;
-import org.springframework.beans.factory.annotation.Autowired;
+import org.apache.commons.lang3.StringUtils;
 import org.springframework.dao.DataAccessException;
-import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
-import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
-import org.springframework.jdbc.core.namedparam.SqlParameterSource;
-import org.springframework.jdbc.support.GeneratedKeyHolder;
-import org.springframework.jdbc.support.KeyHolder;
 import org.springframework.stereotype.Repository;
 
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.*;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Objects;
 import java.util.Optional;
 
-import static com.epam.esm.repository.Parameters.*;
-import static com.epam.esm.repository.builder.CertificateQueries.*;
+import static com.epam.esm.repository.builder.CertificateQueries.DELETE_TAG_FROM_CERTIFICATE;
+import static com.epam.esm.repository.builder.CertificateQueries.INSERT_TAG_TO_CERTIFICATE;
 
 @Repository
 public class CertificateRepositoryImpl implements CertificateRepository {
 
-    private final CertificateMapper rowMapper;
-    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
-
-    @Autowired
-    public CertificateRepositoryImpl(CertificateMapper rowMapper, NamedParameterJdbcTemplate jdbcTemplate) {
-        this.rowMapper = rowMapper;
-        this.namedParameterJdbcTemplate = jdbcTemplate;
+    private static final String TAGS = "tags";
+    private static final String ID = "id";
+    private static final String NAME = "name";
+    private static final String DESCRIPTION = "description";
+    private static final String PARAMETER = "parameter";
+    private static final String PRICE = "price";
+    private static final String DURATION = "duration";
+    private static final String CREATE_DATE = "createDate";
+    private static final String LAST_UPDATE_DATE = "lastUpdateDate";
+    private static final String PARTIAL_STRING = "%%%s%%";
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    public CertificateRepositoryImpl(EntityManager entityManager) {
+        this.entityManager = entityManager;
     }
 
     @Override
     public List<Certificate> find(SearchCriteria searchCriteria) {
-        MapSqlParameterSource parameters = new MapSqlParameterSource();
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Certificate> query = criteriaBuilder.createQuery(Certificate.class);
+        Root<Certificate> certificateRoot = query.from(Certificate.class);

```

Comment:
I've talked about decomposition many times. It is impossible to understand and read this method. Please rewrite

Reply:
Also, this logic, upon request construction, should not be present in the repository. This is a separate responsibility. You are violating the principle of sole responsibility

Reply:
fixed, but not sure if its the right way

Reply:
fixed